### ISSUE TITLE
PyTorch public API cleanup

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#64185]
## 🚀 Feature

We should clean up things that are exposed in the public torch (and other) namespaces but aren't actually intended to be exposed. Examples include torch.batch_norm/torch.group_norm (we only document F.batch_norm and F.group_norm, noticed by @samdow ) and when @soulitzer refactored gradcheck and noted that folks were directly calling gradcheck helper functions.

## Motivation

Tensor extension writers are confused by functions like `torch.batch_norm` because they are overridable but do not appear in the PyTorch docs. Furthermore, these functions can be subtly different from their documented variants (F.batch_norm).

## Pitch

Step 1: Make it so that new functions added to PyTorch are private by default
- We should move all of PyTorch Python source files into a `_src` folder to indicate that they are all private and expose only the APIs that are meant to be exposed. This way, if a dev adds a new helper function inside a file in the _src directory, they are not accidentally exposing it for users to see.
- PyTorch codegenned operators should have the same treatment: they should be a part of a private namespace initially and a user who is adding a new operator must manually expose it.

Step 2: Deprecate all the things we don't want to support
- like torch.batch_norm, torch.group_norm

## Alternatives

N/A

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411

**Comment by zou3519** [comment]
tentatively marking hi-pri because "I would pay someone to fix this"

**Comment by jbschlosser** [comment]
Related: #49480

**Comment by soulitzer** [comment]
@albanD : theres already a script in CI that checks this, but it is silently failing in some cases, so we should review that

**Comment by vadimkantorov** [comment]
Also, may make sense as part of this drive to promote some very frequent functions from torch.nn.functional to torch, e.g. relu, softmax, log_softmax, maybe something else as well

**Comment by albanD** [comment]
Fun fact: all of these are already in `torch.` :D

**Comment by vadimkantorov** [comment]
yes, but within this PR they may get deprecated :)

**Comment by janeyx99** [comment]
@albanD Is this complete with the public api work + tests you and others have completed?

**Comment by albanD** [comment]
I think the "stem the bleeding" part of it is done (with the CI improvements) but the part about "closing the wound" is nowhere done. We would need to empty the allowlists in https://github.com/pytorch/pytorch/blob/main/test/allowlist_for_publicAPI.json and https://github.com/pytorch/pytorch/blob/main/docs/source/conf.py#L97 which have a few thousand of entries each.

**Comment by zou3519** [comment]
Not sure if hi-pri anymore.
